api/src/main/java/org/apache/iceberg/UpdateSnapshotRefs.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import org.apache.iceberg.exceptions.CommitFailedException;
+
+/**
+ * API for snapshot reference evolution.
+ * <p>
+ * When committing, these changes will be applied to the current table metadata.
+ * Commit conflicts will not be resolved and will result in a {@link CommitFailedException}.
+ */
+public interface UpdateSnapshotRefs extends PendingUpdate<Map<String, SnapshotRef>> {
+
+  /**
+   * Assign a tag to a snapshot.
+   *
+   * @param name tag name
+   * @param snapshotId ID of the snapshot to tag
+   * @return this for method chaining
+   * @throws IllegalArgumentException if ref with the same name already exists
+   */
+  UpdateSnapshotRefs tag(String name, long snapshotId);
+
+  /**
+   * Create a branch and set a snapshot as the branch head.
+   *
+   * @param name branch name
+   * @param snapshotId ID of the snapshot to set as the branch head
+   * @return this for method chaining
+   * @throws IllegalArgumentException if ref with the same name already exists
+   */
+  UpdateSnapshotRefs branch(String name, long snapshotId);
```

Comment:
I think last time we ended our discussion here. For branch, we can either create and fail when exist, or do upsert. It seems like there are valid use cases for both, so what about `branch(String name, long snapshotId, boolean replace)`? The same `replace` flag can also be potentially added to `tag` if fit. @rdblue